projects/lua/fuzz_lua.c

```diff
@@ -0,0 +1,154 @@
+  (void)ar;  /* unused arg. */
+  lua_sethook(L, NULL, 0, 0);  /* reset hook */
+  luaL_error(L, "interrupted!");
+}
+
+
+/*
+** Function to be called at a C signal. Because a C signal cannot
+** just change a Lua state (as there is no proper synchronization),
+** this function only sets a hook that, when called, will stop the
+** interpreter.
+*/
+static void laction (int i) {
+  int flag = LUA_MASKCALL | LUA_MASKRET | LUA_MASKLINE | LUA_MASKCOUNT;
+  signal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
+  lua_sethook(globalL, lstop, flag, 1);
+}
+
+
+
+/*
+** Prints an error message, adding the program name in front of it
+** (if present)
+*/
+static void l_message (const char *pname, const char *msg) {
+  if (pname) lua_writestringerror("%s: ", pname);
+  lua_writestringerror("%s\n", msg);
+}
+
+
+/*
+** Check whether 'status' is not OK and, if so, prints the error
+** message on the top of the stack. It assumes that the error object
+** is a string, as it was either generated by Lua or by 'msghandler'.
+*/
+static int report (lua_State *L, int status) {
+  if (status != LUA_OK) {
+    const char *msg = lua_tostring(L, -1);
+    l_message(progname, msg);
+    lua_pop(L, 1);  /* remove message */
+  }
+  return status;
+}
+
+/*
+** Message handler used to run all chunks
+*/
+static int msghandler (lua_State *L) {
+  const char *msg = lua_tostring(L, 1);
+  if (msg == NULL) {  /* is error object not a string? */
+    if (luaL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
+        lua_type(L, -1) == LUA_TSTRING)  /* that produces a string? */
+      return 1;  /* that is the message */
+    else
+      msg = lua_pushfstring(L, "(error object is a %s value)",
+                               luaL_typename(L, 1));
+  }
+  luaL_traceback(L, L, msg, 1);  /* append a standard traceback */
+  return 1;  /* return the traceback */
+}
+
+
+/*
+** Interface to 'lua_pcall', which sets appropriate message function
+** and C-signal handler. Used to run all chunks.
+*/
+static int docall (lua_State *L, int narg, int nres) {
+  int status;
+  int base = lua_gettop(L) - narg;  /* function index */
+  lua_pushcfunction(L, msghandler);  /* push message handler */
+  lua_insert(L, base);  /* put it under function and args */
+  globalL = L;  /* to be available to 'laction' */
+  signal(SIGINT, laction);  /* set C-signal handler */
+  status = lua_pcall(L, narg, nres, base);
+  signal(SIGINT, SIG_DFL); /* reset C-signal handler */
+  lua_remove(L, base);  /* remove message handler from the stack */
+  return status;
+}
+
+
+static int dochunk (lua_State *L, int status) {
+  if (status == LUA_OK) status = docall(L, 0, 0);
+  return report(L, status);
+}
+
+
+/* mark in error messages for incomplete statements */
+#define EOFMARK         "<eof>"
+#define marklen         (sizeof(EOFMARK)/sizeof(char) - 1)
+
+int
+LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    lua_State *L = luaL_newstate();
+    if (L == NULL) {
```

Comment:
Is that actually a good idea to silently ignore state initialization issue?

Reply:
It works well for me, the fuzzer continues runs continuously without issue as well as run on the clusterfuzz servers (from CI output). In either case, if this becomes an issue at any point (e.g. false positives or the fuzzer continuously crashing on the clusterfuzz servers) then we can address it, but I don't see issues now.

Reply:
Problem here is that it won't produce false positives or crashes, but rather silently assume that specific input won't cause issues while is was not tested at all.

Is there any sort of guideline for how to handle errors not related to passed input? It would apply here.

Reply:
The consequences of this happening is not the assumption you state, i.e. in case the environment loading fails then the fuzzers will not assume "this is a good input". What will happen is that these inputs will not trigger a crash in the given execution (since they won't be executed) but they won't be discarded from being tried again later which is likely to happen. We don't have to ensure this sort of resistance to false-negatives in our fuzz drivers.

Reply:
Sound good to me then. Perhaps logging such failure would be a good idea tho.

Reply:
Perfect - I think even logging will be difficult as we won't have access to the logs on the clusterfuzz servers. What we can do is continuously monitor the coverage of the code and if it so happens that we don't get further (because the environment constantly fails to load) then we act. Alternatively we could stop the fuzzer in case the environment won't load, however, I would discourage this because this might sacrifice a lot of other good executions that could happen further in the fuzzing (killing the fuzzer is not recommended).